File: Task3A_B/task3a_b/src/main/java/com/cps2004/FictionBook.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class FictionBook extends Book{

    private String genre;
    private String publicationDate;
    
    // Default constructor
    public FictionBook(){
        super();
        this.genre = "";
        this.publicationDate = "";
    }

    // Parameterized constructor
    public FictionBook(String name, int id, Supplier supplier, int price, int volume, String publisher, int pages, String author, String genre, String publicationDate){
        super(name, id, supplier, price, volume, publisher, pages, author);
        this.genre = genre;
        this.publicationDate = publicationDate;
    }

    // Getters and setters
    public String getGenre() { return genre; }
    public String getPublicationDate() { return publicationDate; }
    public void setGenre(String genre) { this.genre = genre; }
    public void setPublicationDate(String publicationDate) { this.publicationDate = publicationDate; }

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product genre: " + genre);
        System.out.println("Product publication date: " + publicationDate);
    }

    @Override
    // Calculate discounted price
    public double calculate_discount(String month, int quantity) {
        double discountedPrice = getPrice();
        if(month.equals("January") || month.equals("February") || month.equals("March") || month.equals("April") || month.equals("May") || month.equals("June") || month.equals("July") || month.equals("August") || month.equals("September") || month.equals("October") || month.equals("November") || month.equals("December")){
            if(quantity >= 5 && quantity <= 10){
                discountedPrice = getPrice() * 0.9;
            }
            else if(quantity > 10){
                discountedPrice = getPrice() * 0.8;
            }
        }
        return discountedPrice;
    }

}

File: Task1B/fictionbook.cpp
#include "fictionbook.h"

// Default constructor
FictionBook::FictionBook() : Book() {
    this->genre = "";
    this->publicationDate = "";
}

// Parameterised constructor
FictionBook::FictionBook(string name, int price, Supplier* supplier, int volume, int productID, string author, string genre, int pages, string publicationDate) : Book(name, price, supplier, volume, productID, author, genre, pages) {
    this->genre = genre;
    this->publicationDate = publicationDate;
}

// Getters and setters
string FictionBook::getGenre() {
    return this->genre;
}

string FictionBook::getPublicationDate() {
    return this->publicationDate;
}

void FictionBook::setGenre(string genre) {
    this->genre = genre;
}

void FictionBook::setPublicationDate(string publicationDate) {
    this->publicationDate = publicationDate;
}

// Display product information
void FictionBook::display_product_info() {
    Book::display_product_info();
    cout << "Genre: " << this->genre << endl;
    cout << "Publication Date: " << this->publicationDate << endl;
}

// Calculate discount
double FictionBook::calculate_discount(string month, int quantity) {
    if (month == "January" || month == "February" || month == "March" || quantity >= 15) {
        return this->getPrice() * 0.1;
    }
    else {
        return this->getPrice() * 0.05;
    }
}

// Save product information
void FictionBook::save(ostream& out) const {
    out << "fictionbook" << "\n";
    Book::save(out);
    out << genre << "," << publicationDate << "\n";
}

// Load product information
void FictionBook::load(istream& in) {
    Book::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    genre = token;

    getline(ss, token, ',');
    publicationDate = token;
}